### PR TITLE
fix: resolve CVE-2023-45857 in `v0.x` branch

### LINKED ISSUE #6090: Patch `0.x` branch to resolve `CVE-2023-45857` vulnerability
### Is your feature request related to a problem? Please describe.

Barring a migration guide for Axios `1.x`, the latest `0.x` version of Axios should continue to receive updates for non-breaking security vulnerability resolutions.

### Describe the solution you'd like

[CVE-2023-45857](https://www.mend.io/vulnerability-database/CVE-2023-45857) affects versions `0.8.1` thru `1.5.1`. The fix applied to the `1.x` versions in #6028 should be applied to the `0.x` branch as well, and released as a patch.

### Describe alternatives you've considered

_No response_

### Additional context/Screenshots

```bash
Spots that would need mitigation:

https://github.com/axios/axios/blob/880b42e2b8cdd467ce0c6ecd1cf522ef6ef65682/lib/adapters/xhr.js#L143-L156

https://github.com/axios/axios/blob/880b42e2b8cdd467ce0c6ecd1cf522ef6ef65682/test/specs/xsrf.spec.js#L70-L80
```

### PR DESCRIPTION
**Pull request by lnjbr**
Fixes #6090 

Used solutions from #6028 and #6046 to resolve the [CVE-2023-45857 vulnerability](https://www.mend.io/vulnerability-database/CVE-2023-45857) in Axios `0.x` 

Breaking change:
`axios('http://example.com/')` will no longer set an XSRF token by default. To maintain old behavior, a truthy value for `withXSRFToken` must be passed. i.e.  `axios('http://example.com/')` would need to be changed to something akin to:

```
axios('http://example.com/', {
   withXSRFToken: true
});
```

### CODE REVIEW

**Comment by levpachmanov** [comment]
Hey @lnjbr,
We're part of a startup called Seal Security that mitigates software vulnerabilities in older open source versions by backporting/creating standalone security patches - enabling more straightforward remediation in cases like this. We created an axios 0.27.2-sp1 that's vulnerability-free. As with all of our patches, it's open-source and available for free.

If relevant, check out [our GitHub repo](https://github.com/seal-community/patches/tree/main/npm/axios/0.27.2/sp1) if you wish to learn more, or start using our [app](https://app.sealsecurity.io/?ref=ghaxios).

Please feel free to reach us at [info@seal.security](mailto:info@seal.security) if you have any requests/questions.

**Comment by ashnewport** [comment]
Is there a release schedule which might indicate when this vulnerability fix will be released?

**Comment by bmuenzenmeyer** [comment]
@jasonsaayman @DigitalBrainJS sorry for the direct ping, but curious if this vulnerability will be addressed in 0.x?

**Comment by lnjbr** [comment]
> can we check the failing tests? @lnjbr

Yup! All set and ready for the workflows to be re-ran 🙇

**Comment by bmuenzenmeyer** [comment]
@jasonsaayman can the CI be re-ran here?

**Comment by ashnewport** [comment]
How do we trigger the CI here, and assuming green builds, what is the release cadence to main branch? Hoping to get some understand of times to assist with planning.

**Comment by bmuenzenmeyer** [comment]
> How do we trigger the CI here

as a security feature, axios maintiainers have configured the repository to only run CI when they manually kick if off. It's a tactic to ensure forks do not introduce malicious code or try to steal secrets within GH Actions, for example

>  what is the release cadence to main branch

this is correctly targetting the `0.x` branch, and is already on 1.X / main.

**Comment by ashnewport** [comment]
Thanks for the info and running the CI! Much appreciated to get that feedback 👏 

When I said main branch release cadence, what I probably should have asked is: "_when will the changes be merged and released into their respective branch?_", e.g. weekly release on a Monday to the targeted version, or ad hoc, or when certain thresholds are met

**Comment by lnjbr** [comment]
Is there a reason for the [workflow to be running `npm install`](https://github.com/axios/axios/actions/runs/7198852489/job/19880139621?pr=6091#step:4:1) rather than `npm ci` for the Node `12.x` and `14.x` checks when y'all are committing the `package-lock.json` file? I was only able to reproduce the error  in CI locally after regenerating my lock file. All of the dependencies in y'all's `package.json` file are careted and, notably, [deterministic builds as a default were not introduced until NPM 7](https://github.blog/2021-02-02-npm-7-is-now-generally-available/#changes-to-the-lockfile) (i.e. Node 15+). I believe that's causing the discrepancy and, consequentially, the error within the `node_modules` directory.

**TL;DR**: Can I change the install command to `npm ci` for the `12.x` and `14.x` checks or are y'all intentionally avoiding deterministic builds on them?

**Comment by DigitalBrainJS** [comment]
@lnjbr Well, this is an old major branch, no one did CI backports from the 1.x branch...

**Comment by bmuenzenmeyer** [comment]
that looks like permission to change it if you ask me?

**Comment by bmuenzenmeyer** [comment]
@jasonsaayman can you please re-run the CI again?

**Comment by lnjbr** [comment]
Validated that CI for the branch is passing in my repo after replacing `npm install` with `npm ci` in the `ci.yml` file and adding the [`--localTs`](https://github.com/microsoft/dtslint?tab=readme-ov-file#options) option to the `dtslint` command in the `package.json`'s test script 🙇 

[Screenshot of the GH actions page in my repo showing successful builds for the 12.x, 14.x, 16.x, and 18.x build checks in the ci.yml workflow](https://github.com/lnjbr/axios/actions/runs/7451567871)

**Comment by bmuenzenmeyer** [comment]
@DigitalBrainJS can you please re-run the CI here?

**Comment by TerkaPrzemyslaw1897** [comment]
bump!

**Comment by lagatchell** [comment]
> The PR will be released a bit later, because in the `v0.x` branch there is no CI/CD, which means we need to manually release it, like in the wild old days. Backporting CI/CD is unlikely to be advisable for releasing rare security patches. @jasonsaayman can you look into this?

@DigitalBrainJS any update on the release of this?
